### PR TITLE
Store values per-column rather than in _source field

### PR DESCRIPTION
**Pull request by romseygeek**
StoredRowLookup is split into two implementations, a backwards-compatible one
that loads the json-formatted _doc from the _source field, and a new one that constructs
_doc from per-column storage where possible, and from individual stored fields where
no docvalues exist.

Array columns are always serialized to a stored field, as doc-values do not preserve order
or duplicate entries.  Arrays containing objects are handled specially so that any object
sub-fields which are dropped after a row has been inserted can be removed at read time.

Relates to #15548

### CODE REVIEW

File: .idea/inspectionProfiles/Project_Default.xml [hunk 1]
```diff
@@ -1,6 +1,9 @@
 <component name="InspectionProjectProfileManager">
   <profile version="1.0">
     <option name="myName" value="Project Default" />
+    <inspection_tool class="ExtractMethodRecommender" enabled="true" level="WARNING" enabled_by_default="true">
```
Comment (mfussenegger) [inline .idea/inspectionProfiles/Project_Default.xml:4]: Can you revert this?

If this is for some reason useful please extract to a dedicated PR

File: .idea/misc.xml [hunk 2]
```diff
@@ -22,9 +22,8 @@
       </list>
     </option>
   </component>
-  <component name="ProjectRootManager" version="2" languageLevel="JDK_23" default="true" project-jdk-name="temurin-23" project-jdk-type="JavaSDK" />
+  <component name="ProjectRootManager" version="2" languageLevel="JDK_X" default="true" project-jdk-name="temurin-23" project-jdk-type="JavaSDK" />
```
Comment (mfussenegger) [inline .idea/misc.xml:25]: Same here. Please revert. See also https://github.com/crate/crate/pull/16776

File: server/src/main/java/io/crate/execution/dml/GeoShapeIndexer.java [hunk 3]
```diff
@@ -93,7 +90,12 @@ public void indexValue(@NotNull Map<String, Object> value, IndexDocumentBuilder
             SysColumns.FieldNames.NAME,
             name,
             SysColumns.FieldNames.FIELD_TYPE));
+        docBuilder = docBuilder.wrapTranslog(w -> new SidecarTranslogWriter(w, ref.storageIdentLeafName()));
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/dml/GeoShapeIndexer.java:93]: I think this deserves some comment explaining why this is done this way.

File: server/src/main/java/io/crate/types/ArrayType.java [hunk 4]
```diff
@@ -89,12 +92,17 @@ public ArrayType(DataType<T> innerType) {
         StorageSupport innerStorage = innerType.storageSupport();
         if (innerStorage == null) {
             this.storageSupport = null;
+        } else if (ArrayType.unnest(this) instanceof ObjectType || ArrayType.unnest(this) instanceof GeoShapeType) {
+            this.storageSupport = new StorageSupport<T>(innerStorage) {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/types/ArrayType.java:96]: This could also use a comment. Especially the GeoShapeType case.

Although it is currently using a Map for value representation it is not a container type like ObjectType.

File: server/src/main/java/io/crate/types/ArrayType.java [hunk 5]
```diff
@@ -242,6 +250,12 @@ private static <T> List<T> convert(@Nullable Object value,
         }
         if (value instanceof Collection<?> values) {
             return Lists.map(values, convertInner);
+        } else if (value instanceof byte[] b) {
+            try (StreamInput in = new ByteBufferStreamInput(ByteBuffer.wrap(b))) {
+                return in.readList(input -> (T) input.readGenericValue());
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/types/ArrayType.java:258]: So far decoding of the stored format was mostly contained in the column reference implementations. It looks we're now more conflating casts & internal storage - not sure if that's a good idea.

I'd tend to have a `decodeValue` or something like that instead on the `StorageSupport` or something like that. To have some write/read symmetry.
Basically an analog to the ValueIndexer

File: server/src/main/java/io/crate/types/NumericStorage.java [hunk 6]
```diff
@@ -123,6 +124,7 @@ public void indexValue(@NotNull BigDecimal value, IndexDocumentBuilder docBuilde
             if (ref.hasDocValues()) {
                 docBuilder.addField(new SortedNumericDocValuesField(name, longValue));
             } else {
+                docBuilder.addField(new StoredField(name, longValue));
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/types/NumericStorage.java:127]: Why is there no maybeAddStoredField check here?

File: server/src/main/java/io/crate/types/NumericType.java [hunk 7]
```diff
@@ -158,6 +159,10 @@ public BigDecimal sanitizeValue(Object value) {
             BigInteger bigInt = BigInteger.valueOf(longValue);
             return new BigDecimal(bigInt, scale == null ? 0 : scale, mathContext());
         }
+        if (value instanceof byte[] bytes) {
+            var bigInt = NumericUtils.sortableBytesToBigInt(bytes, 0, bytes.length);
+            return new BigDecimal(bigInt, scale == null ? 0 : scale, mathContext());
+        }
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/types/NumericType.java:165]: Doesn't the format depend on the precision? See `NumericStorage.getCollectorExpression`

**Review by mfussenegger — COMMENTED** [review]
Did a first pass and left a couple of comments.

But I skipped a lot of classes I want to have a closer look at it - this is just to give some early feedback.

File: server/src/main/java/io/crate/execution/dml/ObjectIndexer.java [hunk 8]
```diff
@@ -94,35 +99,58 @@ public void indexValue(@NotNull Map<String, Object> value, IndexDocumentBuilder
...
+                columnsToStore.put(k, null);
             }
         });
+        if (docBuilder.maybeAddStoredField()) {
+            if (columnsToStore.isEmpty() == false) {
+                docBuilder.addField(new StoredField(ref.storageIdentLeafName(), toBytes(columnsToStore).toBytesRef()));
+            } else if (value.isEmpty()) {
+                docBuilder.addField(new StoredField(ref.storageIdentLeafName(), toBytes(value).toBytesRef()));
+            }
+        }
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/dml/ObjectIndexer.java:141]: Maybe also add a short comment here explaining why the object needs to store selected key/values separately.

File: server/src/main/java/io/crate/expression/reference/doc/lucene/ColumnFieldVisitor.java [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+    private final Map<String, Field> fields = new HashMap<>();
+    private final Set<ColumnIdent> droppedColumns;
+    private final SourceParser storedSourceParser;
+
+    /**
+     * Creates a new ColumnFieldVisitor for the given table
+     */
+    public ColumnFieldVisitor(DocTableInfo table) {
+        this.droppedColumns
+            = table.droppedColumns().stream().map(Reference::column).collect(Collectors.toUnmodifiableSet());
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/reference/doc/lucene/ColumnFieldVisitor.java:67]: Looking at the usages of `droppedColumns` I think we should consider changing the representation inside `DocTableInfo` to `Map<ColumnIdent, Reference>`. It looks like the `Set<Reference>` representation is almost never a fit.

File: server/src/main/java/io/crate/expression/reference/doc/lucene/ColumnFieldVisitor.java [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+        if (ref.valueType() instanceof ArrayType<?>) {
+            storageName = ArrayIndexer.ARRAY_VALUES_FIELD_PREFIX + storageName;
+            if (ArrayType.unnest(ref.valueType()) instanceof ObjectType o) {
+                fields.put(storageName, new ArrayOfObjectField(ref.valueType(), o, column, storedSourceParser));
+                return;
+            }
+            if (ArrayType.unnest(ref.valueType()) instanceof GeoShapeType) {
+                fields.put(storageName, new GeoShapeField(ref.valueType(), column, storedSourceParser));
+                return;
+            }
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/reference/doc/lucene/ColumnFieldVisitor.java:143]:
```suggestion
            DataType<?> elementType = ArrayType.unnest(ref.valueType());
            if (elementType instanceof ObjectType o) {
                fields.put(storageName, new ArrayOfObjectField(ref.valueType(), o, column, storedSourceParser));
                return;
            }
            if (elementType instanceof GeoShapeType) {
                fields.put(storageName, new GeoShapeField(ref.valueType(), column, storedSourceParser));
                return;
            }
```

File: server/src/main/java/io/crate/expression/reference/doc/lucene/ColumnFieldVisitor.java [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+        }
+    }
+
+    private record GeoShapeField(DataType<?> dataType, ColumnIdent column, SourceParser sourceParser) implements Field {
+        @Override
+        public Object sanitize(Object v) {
+            var map = sourceParser.parse(new BytesArray((byte[]) v), true);
+            return map.values().iterator().next();
+        }
+    }
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/reference/doc/lucene/ColumnFieldVisitor.java:115]: A bit related to my comment regarding having a read/decode in DataType#StorageSupport. Is there a way we could avoid the type specific logic here?

Object/Array special handling is to some degree okay because they're containers, but geo-shape should mostly work like any other value. (We might also change the representation in the future away from a Map, to have concrete properties)

File: server/src/main/java/io/crate/expression/reference/doc/lucene/StoredRowLookup.java [hunk 12]
```diff
@@ -22,96 +22,295 @@
...
+                if (expr instanceof DocCollectorExpression<?>) {
+                    this.fieldsVisitor.registerRef(ref);
+                } else {
+                    var column = ref.toColumn();
+                    if (column.isRoot() == false && column.name().equals(SysColumns.Names.DOC)) {
+                        column = column.shiftRight();
+                    }
+                    expressions.add(new ColumnExpression(expr, column, ref.storageIdent()));
+                }
+            }
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/reference/doc/lucene/StoredRowLookup.java:245]: Similar here - I think the type's StorageSupport should have info on what kind of lookup is necessary instead of doing instanceof checks here

File: server/src/test/java/io/crate/integrationtests/ObjectColumnTest.java [hunk 13]
```diff
@@ -387,13 +387,14 @@ os array(object as (
...
+
+        execute("insert into tbl (id, complex) values (3, {os=[{}, null]})");
         assertThat(response).hasRowCount(1);
 
-        execute("insert into tbl (id, os) values (2, [{}, null, {y=10}, {x=1, y=2}])");
+        execute("insert into tbl (id) values (1)");
         assertThat(response).hasRowCount(1);
 
-        execute("insert into tbl (id, complex) values (3, {os=[{}, null]})");
+        execute("insert into tbl (id, os) values (2, [{}, null, {y=10}, {x=1, y=2}])");
```
Comment (seut) [inline server/src/test/java/io/crate/integrationtests/ObjectColumnTest.java:397]: Are these changes needed/related?

File: server/src/test/java/io/crate/metadata/ColumnIdentTest.java [hunk 14]
```diff
@@ -178,4 +178,5 @@ public void test_replace_paths() {
         assertThat(aaa.replacePrefix(ab)).isEqualTo(aba);
         assertThat(aaa.replacePrefix(aab)).isEqualTo(aab);
     }
+
```
Comment (seut) [inline server/src/test/java/io/crate/metadata/ColumnIdentTest.java:181]: minor: unrelated, can be reverted

File: server/src/testFixtures/java/io/crate/types/DataTypeTestCase.java [hunk 15]
```diff
@@ -92,9 +93,25 @@ public void test_reference_resolver_index_and_docvalues_off() throws Exception {
         doReferenceResolveTest(dataDef.type, dataDef.definition + " INDEX OFF STORAGE WITH (columnstore=false)", dataDef.data.get());
     }
 
-    protected void doReferenceResolveTest(DataType<T> type, String definition, T data) throws Exception {
+    protected boolean supportsArrays() {
+        return true;
+    }
 
-        StorageSupport<? super T> storageSupport = type.storageSupport();
+    public void test_reference_resolver_with_list() throws Exception {
```
Comment (seut) [inline server/src/testFixtures/java/io/crate/types/DataTypeTestCase.java:100]:
```suggestion
    @Test
    public void test_reference_resolver_with_list() throws Exception {
```

File: server/src/testFixtures/java/io/crate/types/DataTypeTestCase.java [hunk 16]
```diff
@@ -147,20 +164,43 @@ protected void doReferenceResolveTest(DataType<T> type, String definition, T dat
...
+    }
+
+    private static String toTypeSignature(DataType<?> dataType) {
+        if (dataType instanceof ObjectType o) {
+            assert o.innerTypes().size() == 1;
+            return "object as (x " + o.innerType("x").getTypeSignature().toString() + ")";
+        } else {
+            return dataType.getTypeSignature().toString();
+        }
+    }
```
Comment (seut) [inline server/src/testFixtures/java/io/crate/types/DataTypeTestCase.java:195]: Is this really needed? `ObjectType.getTypeSignature()` should take care of inner types.

File: server/src/test/java/io/crate/integrationtests/CopyIntegrationTest.java [hunk 17]
```diff
@@ -923,9 +923,9 @@ public void test_copy_excludes_partitioned_values_from_source() throws Exception
                 new Object[] { file.toPath().toUri().toString() }
             );
             execute("refresh table tbl2");
-            execute("SELECT _raw, * FROM tbl2");
+            execute("SELECT _doc, * FROM tbl2");
             assertThat(response).hasRows(
-                "{\"3\":10,\"4\":{}}| 10| {p=1}"
+                "{o={p=1}, x=10}| 10| {p=1}"
```
[the remark this answers is not in the document]
Comment (romseygeek) [reply server/src/test/java/io/crate/integrationtests/CopyIntegrationTest.java:928]: We need to change the `_raw` output entirely - my plan is to get these tests to pass now, and then change `_raw` in a followup where these tests will all need to be updated anyway, and we can use the information they are trying to test to inform what the output format should be.

File: server/src/testFixtures/java/io/crate/types/DataTypeTestCase.java (hunk 16 above)
Comment (romseygeek) [reply server/src/testFixtures/java/io/crate/types/DataTypeTestCase.java:195]: Unfortunately yes - `ObjectType.getTypeSignature().toString` here returns `object(text,"x" byte)` when what we want is `object as (x byte)`.

**Comment by romseygeek** [comment]
Thanks for the feedback @mfussenegger @seut .  I've reworked `ColumnFieldVisitor` so that it now delegates everything to some new methods on `StorageSupport` which is much tidier.  Similarly, whether or not we should load data via expressions or via stored fields is controlled by a new method on DataType.

File: server/src/main/java/io/crate/expression/reference/doc/lucene/ColumnFieldVisitor.java (hunk 9 above)
Comment (romseygeek) [reply server/src/main/java/io/crate/expression/reference/doc/lucene/ColumnFieldVisitor.java:58]: +1, I'll open a separate PR to handle this.

Comment (romseygeek) [reply server/src/main/java/io/crate/expression/reference/doc/lucene/ColumnFieldVisitor.java:58]: Changing this to a Map doesn't really help things, unfortunately.  Generally speaking, callers either want to stream the References or convert them to something that can be accessed via the ColumnIdent, so everything either gets more complex (`droppedColumns()` becomes `droppedColumns.values()`) or stays about the same.  `ColumnFieldVisitor` is the only caller that becomes simpler, so I don't think it's worth it.

File: server/src/main/java/io/crate/types/DataType.java [hunk 18]
```diff
@@ -266,6 +266,13 @@ public final StorageSupport<? super T> storageSupportSafe() {
         return storageSupport;
     }
 
+    /**
+     * @return {@code true} if values should always be loaded from stored fields
+     */
+    public boolean retrieveFromStoredFields() {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/types/DataType.java:272]: I'd say this also belongs to `StorageSupport` because if there's no storage support this info is kinda irrelevant.

File: server/src/main/java/io/crate/expression/reference/doc/lucene/ColumnFieldVisitor.java [hunk 19]
```diff
@@ -0,0 +1,169 @@
...
+    public ColumnFieldVisitor(DocTableInfo table) {
+        this.droppedColumns
+            = table.droppedColumns().stream().map(Reference::column).collect(Collectors.toUnmodifiableSet());
+        this.storedSourceParser = new SourceParser(table);
+    }
+
+    private record Field(DataType<?> dataType, ColumnIdent column) implements Comparable<Field> {
+
+        public Object decode(SourceParser sourceParser, byte[] v) {
+            return dataType.storageSupportSafe().decodeFromBytes(column, sourceParser, v);
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/reference/doc/lucene/ColumnFieldVisitor.java:65]: Probably worth assigning the storageSupport instance to a property of `Field`. Seems like decode is called per doc - and StorageSupport instances can be created on access (e.g. for NumericStorage)

File: server/src/main/java/io/crate/types/BitStringType.java [hunk 20]
```diff
@@ -183,6 +183,9 @@ public BitString sanitizeValue(Object value) {
     public BitString explicitCast(Object value, SessionSettings sessionSettings) throws IllegalArgumentException, ClassCastException {
         // explicit cast is allowed to trim or extend the bitstring
         if (value instanceof String str) {
+            if (str.startsWith("B'")) {
+                str = str.substring(2);
+            }
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/types/BitStringType.java:188]: Should this also go into a decode function?

File: server/src/main/java/io/crate/types/StorageSupport.java [hunk 21]
```diff
@@ -61,6 +66,26 @@ public abstract ValueIndexer<? super T> valueIndexer(
...
+    public Object decodeFromLong(long input) {
+        throw new UnsupportedOperationException("decodeFromLong not supported");
+    }
+
+    /**
+     * Decode a value from an int in a stored field
+     */
+    public Object decodeFromInt(int input) {
+        throw new UnsupportedOperationException("decodeFromInt not supported");
+    }
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/types/StorageSupport.java:88]: Could these return `T` instead of `Object` ? And maybe just call them `decode` - the different overloads should be able to disambiguate.

**Review by mfussenegger — COMMENTED** [review]
Before I forget: If we use the type streamers to serialize/deserialize we need to make sure that the `getVersion()` BWC handling works.

I think for that to work properly we'll have to add `setVersion()` calls on the streams and likely with the index' created-version.

Otherwise than that from what I've seen things look good and fine to me to merge and if necessary iterate further on master, but could also take another closer look on Monday if desired.

File: server/src/test/java/io/crate/integrationtests/CopyIntegrationTest.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (seut) [reply server/src/test/java/io/crate/integrationtests/CopyIntegrationTest.java:928]: I see, can we at least add a comment/todo here to mention this?

File: server/src/main/java/io/crate/execution/dml/ArrayIndexer.java [hunk 22]
```diff
@@ -100,23 +107,26 @@ public static <T> ValueIndexer<T> of(Reference arrayRef, Function<ColumnIdent, R
...
-    private final Reference reference;
+    protected final Streamer<List<T>> innerStreamer;
+    protected final String arrayLengthFieldName;
+    protected final Reference reference;
 
-    private ArrayIndexer(ValueIndexer<T> innerIndexer, Function<ColumnIdent, Reference> getRef, Reference reference) {
+    protected ArrayIndexer(ValueIndexer<T> innerIndexer, Function<ColumnIdent, Reference> getRef, Reference reference) {
         this.innerIndexer = innerIndexer;
-        this.arrayLengthFieldName = toArrayLengthFieldName(reference, getRef);
+        this.innerStreamer = (Streamer<List<T>>) reference.valueType().streamer();
```
Comment (seut) [inline server/src/main/java/io/crate/execution/dml/ArrayIndexer.java:116]: the term  `innerStreamer` confused me at first as we usually refer to the inner element of an array with `inner`.
Maybe lets call it for what it is used?
```suggestion
        this.bytesConverter = (Streamer<List<T>>) reference.valueType().streamer();
```

File: server/src/main/java/io/crate/types/GeoShapeType.java [hunk 23]
```diff
@@ -65,6 +69,20 @@ public ValueIndexer<Map<String, Object>> valueIndexer(RelationName table,
...
+            try (StreamInput in = new ByteBufferStreamInput(ByteBuffer.wrap(bytes))) {
+                return GeoShapeType.INSTANCE.streamer().readValueFrom(in);
+            } catch (IOException ee) {
+                throw new UncheckedIOException(ee);
+            }
+        }
+
+        @Override
+        public boolean retrieveFromStoredFields() {
+            return true;
```
Comment (seut) [inline server/src/main/java/io/crate/types/GeoShapeType.java:84]: maybe worth adding this comment as well?
```suggestion
            // there's a difference in precision between the stored value and the dv value
            // so when we're retrieving for display or exact comparisons, go to the stored
            // value
            return true;
```

**Review by seut — APPROVED** [review]
👍 great! added 2 more minor suggestions

*update*
> Before I forget: If we use the type streamers to serialize/deserialize we need to make sure that the getVersion() BWC handling works.
> I think for that to work properly we'll have to add setVersion() calls on the streams and likely with the index' created-version.

Missed this, good point, we should probably do that before merging.

**Comment by romseygeek** [comment]
I've added version checks to the Streamer encoding/decoding implementations, although it's all future-proofing at the moment as of course everything using an 'old' version format will still be using the previous table format.

**Review by mfussenegger — APPROVED** [review]
